### ISSUE TITLE
Request: domains() list, remove commonName() method

### ISSUE THREAD

**Issue by FGasper** [gknocke/Crypt-PKCS10#4]
subject.CN is deprecated:

http://security.stackexchange.com/questions/69156/using-commonname-for-dns-vs-subjaltname
https://unmitigatedrisk.com/?p=381

At the very least, it’s not useful because it only gives you that one domain, while it’s perfectly common now to have certs (and CSRs) that cover multiple domains.

Particularly the rise of Let’s Encrypt, which doesn’t automatically add the “www.” subdomain to a signed certificate, reinforces the use case of having subjectAltName extensions in CSRs.

This library would be much more useful if it replaced the commonName() method with a domains() method.

I suggest, too, that that method die() if called outside list context, FWIW.

**Comment by FGasper** [comment]
(Incidentally, cPanel has been creating multi-domain CSRs now for a couple years.)

**Comment by tlhackque** [comment]
Certificates are used for people as well as servers.  And much software still uses commonName.  It's not PKCS10's role to force migrations.  Most certificates put one (the "primary") name in commonName, and repeat it (along with any others) in subjectAltName.  That's been the general practice for a long time.

You can get the subjectAltName by calling extensionsValue("subjectAltName").

Note that SANs can contain other things than domains - they contain GeneralNames, which can be DNSnames; but also people, IP addresses, email addresses, URIs, any random OID - and other things.  domains() isn't general enough.  (Note that you can also have domainComponents within an RDN.)

In a shortly forthcoming release (almost complete here in git), list context will return an array.  scalar context will return the first element.  This is compatible with past uses.  This is also true for the subjectName components with dedicated (or auto-vivifying) access methods.

There are access methods for a lot more name components in the new release.  See the doc for details.

I think you can do what you want with the new release, which Gideon will have on CPAN "soon". 

Have a look - and raise an issue (or new reply) if not.

(I'm not the owner, but I did contribute the pending changes.)

**Comment by FGasper** [comment]
> And much software still uses commonName.
Consider: what is the use case for knowing the CN? Why would anyone call that method? They call it not because they care about the CN per se, but to determine the CSR’s domain coverage. But commonName() is wrong for that because it only gives one domain.

I mean, I can admit that *removing* commonName() would be inadvisable given that the module is at a production stage, but as the module is now there is a good likelihood that callers will errantly treat every CSR as covering only one domain. As you imply, we write software for people, and what’s most helpful here is a quick, easy method that gives the CSR’s domain coverage. commonName() only does that partially.

> You can get the subjectAltName by calling extensionsValue("subjectAltName").
That still requires parsing the SAN—which, as you point out, is non-trivial.

I personally think it best to avoid context-sensitive methods, FWIW. It’s bitten my group several times. No biggie either way for me, though; I’m much more concerned with making the domain coverage easy to ascertain.

**Comment by tlhackque** [comment]
Gideon gets the final say; it's his module.

However, you seem to be missing the point that certificates are issued for other purposes than identifying servers.  For example, all my e-mail is digitally signed with a certificate that has an email address, and NO "domain".

The module has to handle all use cases, not just yours.

I don't see what's hard about parsing the subjectAltName extension.

At least with the pending version, for this CSR:

````
Certificate Request:
    Data:
        Version: 0 (0x0)
        Subject: C=AU, ST=Some-State, L=my city, O=Internet Widgits Pty Ltd, OU=Big org, OU=Smaller org, CN=My Name/emailAddress=non
e@no-email.com
        Subject Public Key Info:
            Public Key Algorithm: rsaEncryption
                Public-Key: (1024 bit)
                Modulus:
                    00:9a:cc:e0:1f:65:88:db:88:1c:cd:3c:f0:b0:c1:
                    c7:7c:8b:42:eb:9b:02:75:78:67:71:5f:94:5a:b0:
                    17:b9:4c:c7:25:22:e2:2c:fd:18:4a:4a:a6:dd:35:
                    b4:68:70:f9:91:30:c7:0b:d9:2b:ca:d2:b4:10:ef:
                    f5:19:f4:86:9e:b5:bc:43:1d:29:ee:f6:02:10:1e:
                    e3:6e:77:44:60:02:6e:b0:44:8a:2a:91:42:ab:36:
                    bb:84:6a:df:09:be:7f:11:fc:8b:f0:aa:51:4d:1d:
                    a1:b7:30:ed:74:29:bc:cc:b5:75:91:b9:e1:a5:52:
                    6c:41:24:30:e7:63:9d:89:eb
                Exponent: 65537 (0x10001)
        Attributes:
        Requested Extensions:
            X509v3 Basic Constraints:
                CA:FALSE
            X509v3 Key Usage:
                Digital Signature, Non Repudiation, Key Encipherment
            X509v3 Subject Alternative Name:
                email:noway@none.com, URI:https://fred.example.net, email:someday@nowhere.example.com, DNS:www.example.net, DNS:www.
example.com, DNS:example.net, DNS:example.com
    Signature Algorithm: sha1WithRSAEncryption
         62:a8:3a:9b:c7:80:b1:80:1e:b0:db:a7:07:7a:3d:ac:d5:84:
         30:72:38:62:17:23:e9:f5:e0:5a:ab:a3:ff:ba:1a:f9:5d:55:
         ef:6d:70:f2:e8:ea:9a:dd:6c:7b:47:c8:b9:ba:13:ae:a4:6f:
         21:46:c9:ba:2c:88:f0:fb:9e:e6:27:ca:e5:52:76:58:a0:2c:
         1a:b2:96:02:03:5a:c7:58:03:b7:94:2c:d0:18:31:35:b7:2f:
         88:58:03:ce:32:78:64:55:bc:51:41:78:6d:dc:83:c4:81:36:
         41:d9:7a:1d:6e:4b:48:49:a4:86:83:1a:b4:7b:e7:03:dd:ee:
         c1:27
-----BEGIN CERTIFICATE REQUEST-----
MIICsjCCAhsCAQAwgbMxCzAJBgNVBAYTAkFVMRMwEQYDVQQIDApTb21lLVN0YXRl
MRAwDgYDVQQHDAdteSBjaXR5MSEwHwYDVQQKDBhJbnRlcm5ldCBXaWRnaXRzIFB0
eSBMdGQxEDAOBgNVBAsMB0JpZyBvcmcxFDASBgNVBAsMC1NtYWxsZXIgb3JnMRAw
DgYDVQQDDAdNeSBOYW1lMSAwHgYJKoZIhvcNAQkBFhFub25lQG5vLWVtYWlsLmNv
bTCBnzANBgkqhkiG9w0BAQEFAAOBjQAwgYkCgYEAmszgH2WI24gczTzwsMHHfItC
65sCdXhncV+UWrAXuUzHJSLiLP0YSkqm3TW0aHD5kTDHC9krytK0EO/1GfSGnrW8
Qx0p7vYCEB7jbndEYAJusESKKpFCqza7hGrfCb5/EfyL8KpRTR2htzDtdCm8zLV1
kbnhpVJsQSQw52OdiesCAwEAAaCBvTCBugYJKoZIhvcNAQkOMYGsMIGpMAkGA1Ud
EwQCMAAwCwYDVR0PBAQDAgXgMIGOBgNVHREEgYYwgYOBDm5vd2F5QG5vbmUuY29t
hhhodHRwczovL2ZyZWQuZXhhbXBsZS5uZXSBG3NvbWVkYXlAbm93aGVyZS5leGFt
cGxlLmNvbYIPd3d3LmV4YW1wbGUubmV0gg93d3cuZXhhbXBsZS5jb22CC2V4YW1w
bGUubmV0ggtleGFtcGxlLmNvbTANBgkqhkiG9w0BAQUFAAOBgQBiqDqbx4CxgB6w
26cHej2s1YQwcjhiFyPp9eBaq6P/uhr5XVXvbXDy6Oqa3Wx7R8i5uhOupG8hRsm6
LIjw+57mJ8rlUnZYoCwaspYCA1rHWAO3lCzQGDE1ty+IWAPOMnhkVbxRQXht3IPE
gTZB2XodbktISaSGgxq0e+cD3e7BJw==
-----END CERTIFICATE REQUEST-----
````

We can inspect the subjectAltName as shown below.

You get an array of hashes, one for each component of the name.

````
 p Dumper(  $csr->extensionValue('subjectAltName') )
$VAR1 = [
          {
            'rfc822Name' => 'noway@none.com'
          },
          {
            'uniformResourceIdentifier' => 'https://fred.example.net'
          },
          {
            'rfc822Name' => 'someday@nowhere.example.com'
          },
          {
            'dNSName' => 'www.example.net'
          },
          {
            'dNSName' => 'www.example.com'
          },
          {
            'dNSName' => 'example.net'
          },
          {
            'dNSName' => 'example.com'
          }
        ];
````

Can you explain what you want that's simpler?

**Comment by FGasper** [comment]
> However, you seem to be missing the point that certificates are issued for other purposes than identifying servers.

I certainly understand this! :) What I’m proposing is to require less syntax to answer the question, “what all does this CSR cover?”

By the approach of what’s in here now, we need this logic to answer that question:

```
use List::MoreUtils;
my @domains = uniq(
    ( map { $_->{'dNSName'} // () } @{ $csr->extensionValue('subjectAltName') } ),
);
```
… and that’s assuming we have the ```//``` operator, without which the process becomes more prolix still.

Alternatively, perhaps the following?
```
my @domains = $csr->getNames('dNSName');
```

There is, too, the case of having multiple commonName values. I’ve never seen this, but according to this it’s potentially a thing:
https://secure.comodo.net/api/pdf/latest/AutoApplySSL.pdf

**Comment by tlhackque** [comment]
But then do you also want $csr->getAltNameURIs?  getAltNameEmails?  And all the other possibilities?

And what's with the uniq()?  If the certificate has multiple names, it's not PKCS10's job to compress them.

It feels like you want to put application logic into Crypt::PKCS10.  And I don't think that's right.

Multiple CNs (and OUs and DCs, and...) are correctly handled in the new release.  Yes, they are valid syntax.  In fact, /O /OU /OU /OU is pretty common.

**Comment by FGasper** [comment]
```$csr->getNames($type)``` would solve the general problem, addressing emails, URIs, etc. in addition to domains. I myself think this is a prevalent enough need that it’s sensible to make it require as little syntax as possible.

uniq() is left in from an internal implementation where we combine the CN with SAN/dNSNames and do a uniq() on it. We didn’t trust that the CAs would always include the CN in the SAN. In our own application we always deal with domains, so it’s not a big concern that these two aren’t quite the same kind of thing.

If Gideon would rather have people do map/grep to get all of the covered subject names, I guess we’ll just subclass locally, but I’d be a bit more afraid that John Q. Caller will just continue calling commonName() in scalar context.

**Comment by FGasper** [comment]
Maybe ```$csr->getAltNames($type)```.

**Comment by tlhackque** [comment]
It would have to be getAltNames, as there are lots of names in a CSR :-)

Note that multiple CNs has been known to happen with Marriage/Divorce, Professional names, English/Chinese names.  So you're still going to get an array back.  It can have multiples of any type.

John Q Caller better understand the contents of CSRs if he's writing a PKI.  Or he has bigger problems.

It wouldn't be hard to implement getAltNames - though it would add unique code.  (Currently, all access methods share common code.) I still think it's an application layer issue.  But I could be had.

Let's let Gideon weigh in.  If he agrees with you, I could be persuaded to implement it - I've been in the code recently.

**Comment by FGasper** [comment]
On 8 Jan 2016 7:23 PM, Timothe Litt wrote:
>
> Note that multiple CNs has been known to happen with Marriage/Divorce,
> Professional names, English/Chinese names. So you're still going to get
> an array back. It can have multiples of any type.

Interesting … I myself have never seen it. I thought that was a pattern 
that was considered and rejected when certs were being expanded to cover 
multiple subject names.

>
> John Q Caller better understand the contents of CSRs if he's writing a
> PKI. Or he has bigger problems.

Right now the SYNOPSIS seems misleading. The Comodo link above talks 
about “primary domains”; there still seems a widespread concept that a 
cert covers one name.

>
> It wouldn't be hard to implement getAltNames - though it would add
> unique code. (Currently, all access methods share common code.) I still
> think it's an application layer issue. But I could be had.

I do see your point on that and am grateful for your thoughtful 
responses tonight.

>
> Let's let Gideon weigh in. If he agrees with you, I could be persuaded
> to implement it - I've been in the code recently.
>

Cool--cheers, and have a great weekend! :)

-Felipe Gasper
Houston, TX

**Comment by tlhackque** [comment]
>>Interesting … I myself have never seen it. I thought that was a pattern
that was considered and rejected when certs were being expanded to cover
multiple subject names. ... about “primary domains”; there still seems a widespread concept that a
cert covers one name.

The cases I mention aren't multiple subjects.  They're the same subject (person), who happens to have multiple names.

For example, Miss Jane Elizabeth Maiden becomes Mrs. Jane  Maiden Married, but is known professionally as Judge Liz Maiden.

This might be /C:Wierd/ST:TX/L:Houston/O=Courts/OU=Superior/OU=Childrens/OU=District 7/CN=Jane Elizabeth Maiden/CN=Jane Maiden Married/CN=Liz Maiden

I'm not recommending this; I think these would be better as SANs.  But I've certainly seen it in LDAP and X.400 directories. People do a wildcard lookup for an org unit, then grep thru names looking for a match on any CN.  This is somewhat easier to code than retrieving a list from a SAN - where you STILL have to grep for the desired CN in each one.  And it worked before the SAN was defined.  So code had to handle this form as well as the SAN.  (So why do both, given that both issuers and consumers have to be backward-compatible?)  Laziness is no excuse.  But it is a reason :-(

Still, the certificate represents one subject - the same human.

This is somewhat different from the common server (domain name case).  In that case, multiple subjects are being covered by one certificate.  (Which is a perversion of the original idea of certificate.)  Each subject is, for example, a (possibly virtual) webserver.  And while www.example.com and example.com may serve the same content, it's not uncommon for one certificate to cover www.example.info and shopping.example.info - where each has it's own server and resides on its own machine.  And the example.com may be used for SMTP/IMAP servers too.  People do this to save money on certificates - and for failover and load sharing cases.

So, yes, a certificate was intended to represent a single subject (e.g. person or server), possibly with multiple names.  

But it's evolved due to technical (apache virtual server and SSL prior to SNI) and economic (outrageous cost of commercial certificates) reasons to where a certificate is used to represent multiple subjects who share some authentication domain.  That's not the original model, but it is real life.  It's not ideal from a security or architectural point of view.  And it does create the confusion that you mentioned.  

Is a certificate an authenticator of one subject with alternate names?  Or is it a an authenticator of multiple subjects who happen to share a common authentication domain?

The answer depends on trade-offs that we can't legislate here.  At this point, neither model is 'right' in an absolute sense.

Finally, the "primary name" idea that you mention comes from the late introduction of SANs.  For webservers, CN in the subject was all there was.  

When SAN was introduced, you couldn't count on clients looking at it.  And those that did sometimes only looked at the first value.  Crypt:::PKCS10 suffered from this until recently.  (unfortunately, still somewhat true of clients today.  Not so much browsers, which seem to have caught up.  But non-http protocols often don't talk about SANs - or even wildcards.)  So the advice was to put the most important, or "primary" name in the CN as it will work with clients and protocols that don't know about the newfangled SAN extension.  And put secondary names in the SAN (after the "primary") , to minimize the business impact of clients that don't look (or look far enough into) the SAN.

SAN isn't exactly "newfangled" any more.  But the vocabulary persists.

**Comment by FGasper** [comment]
I actually did say “multiple subject names”. ;-) But, thank you for the interesting thoughts and history!

**Comment by tlhackque** [comment]
I do agree that the SYNOPSIS line that references commonName should be deleted.  A better choice would be:
`my $subject = $decoded->subject;`

This code trivially does what you suggest.  It could go in Crypt::PKCS10, or you can place and call it in your application.  It has no knowledge of Crypt::PKCS10 internals, which is why I think it's an application layer function.  It works with Perl 5.8.8 - no need for //.

It returns the values of the given type (or the first value).  Or empty array/undef if the extension is missing.

````
# $type is any one of (The ones marked ## are the most common/useful):
# otherName 
## rfc822Name 
## dNSName
# x400Address
# directoryName
# ediPartyName 
## uniformResourceIdentifier  
## iPAddress, 
# registeredID                     

sub subjectAltName {
    my $self = shift;
    my( $type ) = @_;

    my $san = $self->extensionValue( 'subjectAltName' );
    unless( defined $san && defined $type ) {
        return () if( wantarray );
        return undef;
    }

    my $result = [ map { $_->{$type} } grep { exists $_->{$type} } @$san ];

    return @$result if( wantarray );
    return $result->[0];
}
````
Usage (shown in Crypt::PKCS10 with the above test cert):

````
DB<1> x $csr->subjectAltName('dNSName' )
0  'www.example.net'
1  'www.example.com'
2  'example.net'
3  'example.com'
DB<2> x scalar $csr->subjectAltName('dNSName' )
0  'www.example.net'
DB<3> x scalar $csr->subjectAltName('rfc822Name' )
0  'noway@none.com'
DB<4> x  $csr->subjectAltName('rfc822Name' )
0  'noway@none.com'
1  'someday@nowhere.example.com'
DB<5> x  $csr->subjectAltName('uniformResourceIdentifier' )
0  'https://fred.example.net'
DB<6> x  $csr->subjectAltName('uniformResourceIdentifierX' )
  empty array
DB<7> x  scalar $csr->subjectAltName('uniformResourceIdentifierX' )
0  undef
DB<8>
````

If not in Crypt::PKCS10, in your application, just call as
` subjectAltName($decoded, $type)`

**Comment by FGasper** [comment]
A reference to subject() would have the same problem as the present SYNOPSIS: complicating the task of finding out what is, along with the key, arguably *the* most important information in the CSR: how we label the subject.

I surmise that most people who call into this module don’t know that certs can have multiple CNs, don’t know that CN is technically deprecated but still everywhere, and generally aren’t very familiar with the format. They likely would, upon seeing the SYNOPSIS, run that through Data::Dumper, see the domain in commonName, and stop there.

The important piece of information, again, is not the subject.CN per se, but *how we label the subject*. The mechanics of how the subject.CN and SAN interact to answer that question is, quite arguably, not application-level information, but an encoding detail. The application should not care about how the information is stored in the CSR, but forcing callers to distinguish between the subject.CN and the SAN does precisely that.

Even if ```subjectAltName($type)``` is indeed just syntactic sugar, it’s sugar that I suspect almost every application would find convenient. I’m normally all for not creating multiple ways of doing things (heh, making me an oddball among Perl programmers!), but IMO it’s quite justified here.

**Comment by tlhackque** [comment]
You can find an implementation in my fork (tlhackque/Crypt-PKCS10).  It won't stay there unless Gideon agrees on the value.  (In which case he'll merge it into his master.)  But for now you can play with it.

Enjoy.

**Comment by FGasper** [comment]
I was about to create a pull request. :)

**Comment by tlhackque** [comment]
I hope that your assessment of the user is wrong.  if it's correct, they're in far more trouble than this bit of syntax can possibly fix.  If you create certificates without understanding the semantics, you're doomed to a lot of failures until you do.

There is a universal tendency to believe that "my" application is representative of all applications (for any value of "my").  Sorry, it isn't true.  (Even when "my" really is mine.)

My application doesn't benefit from the function.  For users, I, the SAN is created from trustworthy data.  For CSRs from devices, it's reviewed manually.  In either case, either the whole thing is copied to the certificate, or the request is rejected.

 In any case, it is not the job of Crypt::PKCS10 to educate or influence its users about the semantics of certificates.

We'll see if Gideon is persuaded by your argument.  I'm going back to my real project.

**Comment by mbartosch** [comment]
I am working for the OpenXPKI project, an Open Source PKI solution. Crypt::PKCS10 will be used in future versions of the trustcenter software to parse incoming requests.
For the CA side of request parsing it is absolutely essential to be able to extract the Subject as well as SubjectAlternativeNames from the CSR.
I disagree with the notion that the interface to the parsing module should "simplify" things by non-standard interpretation. This is a technical module, using it properly is the responsibility of the user.

Regarding the purpose of the PKCS#10 components I strongly suggest to stick to RFC5280 and not try to be too clever about "modern interpretations" of the standard.
The subject of a CSR and the resulting certificate shall identify the holder of the certificate. In the context of humans this may be the name of the person (or names in rare cases), for systems in my opinion by far the most useful interpretation is to assign the FQDN of the system where the certificate is installed (!) and NOT a symbolic name (DNS Alias for a virtual host). Then all virtual host FQDNs should be listed as DNS SubjectAltNames, *including* the FQDN of the primary host.

(The RFC for TLS basically specifies that a relying party shall consider the DNS SubjectAlternativeNames if they exist and completely ignore the Subject if they exist, so the Subject is totally irrelevant in the context of TLS IFFF at least one SAN exists. Only in legacy cases where no DNS SAN exists client implementations fall back to evaluating the Subject. This is something many people do not know, leading to all kinds of bad implementations. Seen them all...)

Bottom line is, I mostly agree with tihackque's view on things. 
Regarding of providing non-standard API accessors I strongly suggest to locally subclass or wrap this module.
I recommend not to add non-standard accessors, as this will only contribute to more broken client software.

**Comment by tlhackque** [comment]
Thanks Martin.

You may want to weigh in on another pull request of mine; this morning I realized that IP addresses were coming out as binary blobs rather than text.  Text is the 'right' answer, but it does represent an API change.(1)  See #6 for the rationale & let us know if you have a strong feeling.

I think that Crypt::PKC10's role is to provide means to extract whatever is in the CSR; policy of what to do with it is up to the application.  My recent changes have been to make it possible to access more of the CSR contents.  Do let us know if I missed anything that you need.

subjectAltName is a convenience wrapper; it's not entirely bad, though it does clutter the API. I don't think it helps or hurts the cause of better client software.  The plus side is that it makes selecting components of the SAN easier.  I suppose that makes it marginally easier to write better client software.

As I commented on Filipe's pull request: I'm not violently opposed; I simply am not anxious to push application layer functions into Crypt::PKCS10.  This one is borderline.

(1) There are lots of reasons that putting IP addresses into a SAN is a bad idea, so it's uncommon.  But it is done.

**Comment by FGasper** [comment]
Far be it from me to presume what is of relevance to a CA in parsing CSRs … and, indeed, that application is likely the most immediate use case for such.

Thank you both for your input. If nothing else, I’d be happy to send in a pull request to expand on the SYNOPSIS a bit. What got me, honestly, was seeing commonName() as the only example invocation. That *is* deprecated, so I don’t think I’m being unreasonable to suggest it would be better to illustrate calls like ```subject()```, ```extensionValue()```, etc.

Maybe even an EXAMPLES section that specifically shows how to grab all of the domains for a “garden-variety” server-name-based CSR.

**Comment by tlhackque** [comment]
I agree that the SYNOPSIS needs work.  By convention, it should have a list of all the API functions, which it certainly doesn't.  I do think that what you did to the synopsis is positive - modulo where subjectAltName lives.

I documented what I did further down in the pod, but didn't upgrade the doc overall.  If someone has energy & Gideon approves, it could use a rewrite.  (Certainly, the place where it says "and just run Dumper to see what you get" is deficient!

Examples would be a good subdirectory.  That way they're executable, and not space-limited.

I expect that if you contribute some, they'll be favorably received.

I'm afraid that redoing the doc is beyond what I have cycles for.  I keep thinking I'm done & then find functional issues...

**Comment by gknocke** [comment]
First of all thanks for your thoughts and proposals.
When I started developing the tool I simplified some datastructures only to find out that I failed to address some use-cases - multi-valued RDNs, for example. Thats why I got very careful with interpreting data structures. I would be more happy with having the data structure the way it is, without interpreting it with a certain use-case in mind.

The documentation is insufficient: Unfortunately, I currently have very limmited time to improve it. But if someone want's to make improvements, I am very happy about it!
